pkg/controller/dnsrecord/actuator.go

```diff
@@ -144,8 +145,21 @@ func (a *actuator) Migrate(ctx context.Context, dns *extensionsv1alpha1.DNSRecor
 
 func (a *actuator) getDomainName(ctx context.Context, dns *extensionsv1alpha1.DNSRecord, dnsClient alicloudclient.DNS) (string, error) {
 	switch {
-	case dns.Spec.Zone != nil && *dns.Spec.Zone != "":
-		return *dns.Spec.Zone, nil
+	case dns.Spec.Zone != nil && *dns.Spec.Zone != "" && (dns.Status.Zone == nil || *dns.Status.Zone == "" || !zoneMatchesDomainName(*dns.Spec.Zone, *dns.Status.Zone)):
+		if isDomainName(*dns.Spec.Zone) {
+			return *dns.Spec.Zone, nil
+		}
+		// The value specified in dns.Spec.Zone is not a domain name, so assume it's a domain id,
+		// and try to determine the domain name by getting the name of the domain with this id
+		domainName, err := dnsClient.GetDomainName(ctx, *dns.Spec.Zone)
```

Comment:
I have concerns about the case if the value specified in the dns.Spec.Zone is a domain id.
In this case you have a `DescribeDomains` request for the first reconciliation of every single `DNSRecord`.
Wouldn't it be better if the result is cached? The domains will only change very infrequently. And if I count correctly there can be four `DNSRecords` for a cluster.

Reply:
Yes, you are right, I deliberately wanted to avoid caching in `DNSRecord` controllers to keep them simple, but I guess in this case I should make an exception, since the main mechanism to avoid getting the domains (namely specifying them explicitly) is essentially broken if domain ids, and not domain names, are specified.

Reply:
OK, I added caching for domains now. It uses an expiring cache with a TTL of 1 hour, so `DescribeDomains` results will be cached (per `accessKeyId`) for 1 hour before they expire and have to be fetched again. Do you consider this good enough, or would you propose a different caching strategy or TTL value?

Reply:
@MartinWeindel  I don't remember if you implement DNS client rate limit before. In future, I would also suggest cache client too and enforce ratelimit config for each client.

Reply:
@jia-jerry Yes we use a client-side rate limiter for all providers in external-dns-management. The [TokenBucketRateLimiter](https://github.com/kubernetes/client-go/blob/master/util/flowcontrol/throttle.go#L63) is used for this purpose.

Reply:
Such a rate limiter was recently added to the AWS provider extension, see gardener/gardener-extension-provider-aws/pull/425, after we experienced actual issues with exceeding the route53 rate limits on some of our landscapes. We could something similar here as well (in a different PR). What are the Alicloud rate limits? Do you expect that they could be exceeded, have we actually experienced such issues in the past?

Reply:
We encountered it before https://github.com/gardener/external-dns-management/pull/77.

Reply:
@stoyanr, would you please take care of https://github.com/gardener/gardener-extension-provider-alicloud/issues/387?

Reply:
I will, in a separate PR, as mentioned.